common/dev/ltc4282.c

```diff
@@ -1,200 +1,283 @@
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	ltc4282_init_arg *init_arg =
+		(ltc4282_init_arg *)sensor_config[sensor_config_index_map[sensor_num]].init_args;
+
+	if (!init_arg->is_init) {
+		printf("[%s], device isn't initialized\n", __func__);
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	if (!init_arg->r_sense_mohm) {
+		printf("%s, Rsense hasn't given\n", __func__);
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	float rsense_mohm = init_arg->r_sense_mohm / 1000;
+	float full_scale_output_voltage;
+	for (uint8_t cnt = 0; cnt <= ARRAY_SIZE(FOLDBACK_MODE_TABLE); cnt++) {
+		if (init_arg->ilim_adjust.fields.foldback_mode ==
+		    FOLDBACK_MODE_TABLE[cnt].v_range_selection) {
+			full_scale_output_voltage =
+				FOLDBACK_MODE_TABLE[cnt].full_scale_output_voltage;
+			break;
+		}
+		if (cnt == ARRAY_SIZE(FOLDBACK_MODE_TABLE)) {
+			printf("[%s] Unknown voltage range, the foldback mode 0x%x\n", __func__,
+			       init_arg->ilim_adjust.fields.foldback_mode);
+			return SENSOR_UNSPECIFIED_ERROR;
+		}
+	}
+
+	uint8_t retry = 5;
+	double val = 0;
+	I2C_MSG msg = { 0 };
+
+	sensor_cfg *cfg = &sensor_config[sensor_config_index_map[sensor_num]];
+
+	msg.bus = cfg->port;
+	msg.target_addr = cfg->target_addr;
+	msg.tx_len = 1;
+	msg.data[0] = cfg->offset;
+	msg.rx_len = 2;
+
+	if (cfg->offset != LTC4282_ENERGY_OFFSET) {
+		if (i2c_master_read(&msg, retry) != 0)
+			return SENSOR_FAIL_TO_ACCESS;
+	}
+	// Refer to LTC4282 datasheet page 23.
+	switch (cfg->offset) {
+	case LTC4282_VSENSE_OFFSET:
+		val = (((msg.data[0] << 8) | msg.data[1]) * 0.04 / 65535 / rsense_mohm);
+		break;
+	case LTC4282_POWER_OFFSET:
+		val = (((msg.data[0] << 8) | msg.data[1]) * full_scale_output_voltage * 0.04 *
+		       65536 / 65535 / 65535 / rsense_mohm);
+		break;
+	case LTC4282_VSOURCE_OFFSET:
+		val = (((msg.data[0] << 8) | msg.data[1]) * full_scale_output_voltage / 65535);
+		break;
+	case LTC4282_ENERGY_OFFSET:
+		if (ltc4282_read_ein(&msg, &val, retry) < 0) {
+			return SENSOR_FAIL_TO_ACCESS;
+		}
+		val = (val * 16.64 * 0.04 * 256 / 65535 / 65535 / rsense_mohm);
+		break;
+	default:
+		printf("Invalid sensor 0x%x offset 0x%x\n", sensor_num, cfg->offset);
+		return SENSOR_NOT_FOUND;
+	}
+
+	sensor_val *sval = (sensor_val *)reading;
+	memset(sval, 0, sizeof(*sval));
+
+	sval->integer = (int)val & 0xFFFF;
+	sval->fraction = (val - sval->integer) * 1000;
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t ltc4282_init(uint8_t sensor_num)
+{
+	if (sensor_num > SENSOR_NUM_MAX) {
+		return SENSOR_INIT_UNSPECIFIED_ERROR;
+	}
+
+	I2C_MSG msg;
+	uint8_t retry = 5;
+
+	msg.bus = sensor_config[sensor_config_index_map[sensor_num]].port;
+	msg.target_addr = sensor_config[sensor_config_index_map[sensor_num]].target_addr;
+
+	ltc4282_init_arg *init_args =
+		(ltc4282_init_arg *)sensor_config[sensor_config_index_map[sensor_num]].init_args;
+
+	if (init_args->is_init) {
+		goto exit;
+	}
+
+	if (init_args->is_register_setting_needed == 0x0) {
+		goto init_param;
+	}
+
+	int bit = 0;
+	do {
```

Comment:
Please convert this to a for loop for safety.

Reply:
The code is forced updated.
Please review it again.
Thanks.